lib/attrsets.nix

```diff
@@ -823,6 +823,57 @@ rec {
         ) a (attrNames n)
     ) {} list_of_attrs;
 
+  /**
+    Flatten an attribute set to the bottom name and value pairs as
+    one singular final attribute set, stripping everything else
+    above.
+
+    In a few words, this does what `lib.lists.flatten` does, but for
+    attribute sets, rather than lists.
+
+
+    # Inputs
+
+    `attrs`
+
+    : The top-level attribute set to flatten.
+
+    # Type
+
+    ```
+    flattenAttrs :: AttrSet -> AttrSet
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.flattenAttrs` usage example
+
+    ```nix
+    flattenAttrs { a.b.c.d = "e"; one.two.three.four.five = 6; foo.bar = null; }
+    => { d = "e"; five = 6; bar = null; }
+    ```
+
+    :::
+  */
+  flattenAttrs = attrs:
```

Comment:
I think we should choose a different name.

When reading flattenAttrs i would think of something like:

```nix
    flattenAttrs { a.b.c.d = "e"; }
    => { "a.b.c.d" = "e"; }
```

It **only** flattens. 

But your function removes information. Going in the other way with something like `expandAttrs` as the inverse of your function is impossible.

Reply:
> `"a.b.c.d"`

This still removes information, because `concatStringsSep`, the implied behavior, is not injective. Counterexample:

Both of the attribute paths `a.b.c.d` and `"a.b".c.d` map to the attribute name `"a.b.c.d"`.

Injectivity is often not a requirement, but I think you're right that a `flatten` operation should generally preserve all information except the nesting structure.

How to do that here is not obvious, but that can be solved by adding a parameter to make that explicit.

I would suggest to make the type `(List String -> String) -> Attrs -> Attrs`, where the first parameter decides what to do with the attribute path.

Then at least the call sites have a hint what happens, and it makes the function more versatile.
Examples:

- `flattenAttrs (concatStringsSep ".") foo`
- `flattenAttrs last foo`

A rename to something like `flattenAttrPaths` would make it possible to guess the behavior without reading the docs, although the flattening happens to the attrsets, and not the paths, so that would be `joinAttrPaths` or `joinAttrsRecursive` to match another pattern that we have in `lib.attrsets`. (Maybe not `joinAttrs`, which I've proposed [here](https://github.com/NixOS/nixpkgs/pull/201527#discussion_r1025207834))

Reply:
@roberth this is a duplicate implementation/feature-set of https://github.com/NixOS/nixpkgs/pull/221608

I think we should merge `collect'` which is the smallest common denominator here.